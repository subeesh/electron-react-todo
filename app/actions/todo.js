import * as types from "../constants/todoActionTypes";

export const addTodo = text => ({ type: types.ADD_TODO, payload: text });
export const deleteTodo = id => ({ type: types.DELETE_TODO, payload: id });
export const editTodo = (id, text) => ({
  type: types.EDIT_TODO,
  payload: { id, text }
});
export const completeTodo = id => ({ type: types.COMPLETE_TODO, payload: id });
export const completeAll = () => ({ type: types.COMPLETE_ALL, payload: null });
export const clearCompleted = () => ({
  type: types.CLEAR_COMPLETED,
  payload: null
});
