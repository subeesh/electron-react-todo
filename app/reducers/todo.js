import {
  ADD_TODO,
  DELETE_TODO,
  EDIT_TODO,
  COMPLETE_TODO,
  COMPLETE_ALL,
  CLEAR_COMPLETED
} from "../constants/todoActionTypes";

const initialState = [];

export default function todos(state = initialState, action) {
  switch (action.type) {
    case ADD_TODO:
      return [
        {
          id: state.reduce((maxId, todo) => Math.max(todo.id, maxId), -1) + 1,
          completed: false,
          text: action.payload
        },
        ...state
      ];

    case DELETE_TODO:
      return state.filter(todo => todo.id !== action.payload);

    case EDIT_TODO:
      return state.map(
        todo =>
          todo.id === action.payload.id
            ? { ...todo, text: action.payload.text }
            : todo
      );

    case COMPLETE_TODO:
      return state.map(
        todo =>
          todo.id === action.payload
            ? { ...todo, completed: !todo.completed }
            : todo
      );

    case COMPLETE_ALL:
      const areAllMarked = state.every(todo => todo.completed);
      return state.map(todo => ({
        ...todo,
        completed: !areAllMarked
      }));

    case CLEAR_COMPLETED:
      return state.filter(todo => todo.completed === false);

    default:
      return state;
  }
}
