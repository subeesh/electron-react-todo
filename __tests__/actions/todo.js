import * as types from "../../app/constants/todoActionTypes";
import * as actions from "../../app/actions/todo";

test("addTodo creates ADD_TODO action", () => {
  expect(actions.addTodo("Learn React")).toMatchSnapshot();
});

test("deleteTodo creates DELETE_TODO action", () => {
  expect(actions.deleteTodo(1)).toMatchSnapshot();
});

test("editTodo creates EDIT_TODO action", () => {
  expect(actions.editTodo(1, "Learn Redux")).toMatchSnapshot();
});

test("completeTodo creates COMPLETE_TODO action", () => {
  expect(actions.completeTodo(1)).toMatchSnapshot();
});

test("completeAll creates COMPLETE_ALL action", () => {
  expect(actions.completeAll()).toEqual({
    payload: null,
    type: "COMPLETE_ALL"
  });
});

test("clearCompleted creates CLEAR_COMPLETED action", () => {
  expect(actions.clearCompleted()).toEqual({
    payload: null,
    type: "CLEAR_COMPLETED"
  });
});
