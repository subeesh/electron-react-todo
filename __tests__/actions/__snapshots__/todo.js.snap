// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addTodo creates ADD_TODO action 1`] = `
Object {
  "payload": "Learn React",
  "type": "ADD_TODO",
}
`;

exports[`completeTodo creates COMPLETE_TODO action 1`] = `
Object {
  "payload": 1,
  "type": "COMPLETE_TODO",
}
`;

exports[`deleteTodo creates DELETE_TODO action 1`] = `
Object {
  "payload": 1,
  "type": "DELETE_TODO",
}
`;

exports[`editTodo creates EDIT_TODO action 1`] = `
Object {
  "payload": Object {
    "id": 1,
    "text": "Learn Redux",
  },
  "type": "EDIT_TODO",
}
`;
